### PR TITLE
Feature/multiplexer Multiplex addresses (WIP)

### PR DESCRIPTION
**Pull request by Recmo**
Turns a single address role (like `owner`) into a multi-address one without modifying or redeploying the original contract.

It's labeled WIP because the tests are not finished yet. `yarn test` fails to run any tests on master ATM.

### CODE REVIEW

File: contracts/Multiplexer.sol [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+contract Multiplexer {
+
+    ////////////////
+    // State
+    ////////////////
+
+    /// @dev The current owner. We can not inherit from Ownable here because
+    ///     the public functions may collide with public functions on the target
+    ///     contract that we want to forward too.
+    address private owner;
```
Comment (SylTi) [inline contracts/Multiplexer.sol]: You should use Ownable.sol

File: contracts/Multiplexer.sol [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+    /// @dev The candidate owner. For safety reasons, we do a
+    //      transfer-accept handshake to prevent the ownership
+    //      from being lost. See `owner`.
+    address private ownerCandidate;
```
Comment (SylTi) [inline contracts/Multiplexer.sol]: I like that, but you should extend Ownable.sol (OwnableWithHandShake?) and inherit from the resulting object

File: contracts/Multiplexer.sol [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    event ManagerRemoved(address manager);
+
+    ////////////////
+    // Modifiers
+    ////////////////
+
+    /// @dev Only owner is allowed. We reimplement it here instead of
+    ///     inheriting from Ownable so we can work arround the name
+    ///     collision.
+    modifier onlyOwner() {
```
Comment (SylTi) [inline contracts/Multiplexer.sol]: Ownable.sol

File: contracts/Multiplexer.sol [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    ///     target contract. In the example, the target
+    ///     contract is the existing `HotWallet` instance.
+    address private target;
+
+    /// @dev Mapping for quick lookup if a particular address belongs to a
+    ///     manager.
+    mapping (address => bool) private isManager;
+
+    /// @dev List for enumerating all current managers.
+    address[] managersList;
```
Comment (SylTi) [inline contracts/Multiplexer.sol]: Maybe you should extract that into a SharedOwnership (not sure about the name) and make the multiplexer inherit from it?

File: contracts/Multiplexer.sol [hunk 5]
```diff
@@ -0,0 +1,316 @@
+pragma solidity ^0.4.10;
+
+/// @title Address multiplexer
```
Comment (SylTi) [inline contracts/Multiplexer.sol]: why is all of this commented instead of removed?

**Comment by ajsantander** [comment]
@Recmo any luck with the tests?

**Comment by federicobond** [comment]
Thank you for your contribution @Recmo! I was confused by the owner / manager terminology at first. May I suggest we use the term `operator` as in ERC777? We could then have `authorizeOperator(address)` and `revokeOperator(address)` functions.

Other than that, it would be great to do full return value proxying like it's done [here](https://github.com/zeppelinos/labs/blob/b9f09ac/upgradeability_using_inherited_storage/contracts/Proxy.sol) but maybe there is a reason why it's not yet included in Zeppelin core.

**Comment by nventuro** [comment]
Closing due to staleness. Note that this sort of behaviour can be easily implemented using our `Roles` contracts.

**Pull request closed without merging**